src/image-pause-service.ts

```diff
@@ -0,0 +1,144 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { CONFIG, PROMOTION_CONFIG, inPromotionMode } from './config';
+import { GcsApi } from './gcs-api';
+import { GoogleAdsApi } from './google-ads-api';
+import { Triggerable } from './triggerable';
+
+export class ImagePauseService extends Triggerable {
+  private readonly _gcsApi;
+  private readonly _googleAdsApi;
+  private readonly _config;
+  private readonly _isPromotionMode: boolean;
+
+  constructor(isPromotionMode: boolean) {
+    super();
+    this._isPromotionMode = isPromotionMode;
+    this._config = this._isPromotionMode ? PROMOTION_CONFIG : CONFIG;
+    this._gcsApi = new GcsApi(this._config['GCS Bucket']);
+    this._googleAdsApi = new GoogleAdsApi(
+      this._config['Ads API Key'],
+      this._config['Manager ID'],
+      this._config['Account ID']
+    );
+  }
+
+  run() {
+    this.deleteTrigger();
+    const adGroups = this._googleAdsApi.getAdGroups();
+    const lastProcessedKey = this._isPromotionMode
+      ? 'lastPromotionImagePauseProcessedAdGroupId'
+      : 'lastImagePauseProcessedAdGroupId';
+    const lastImagePauseProcessedAdGroupId =
+      PropertiesService.getScriptProperties().getProperty(lastProcessedKey);
+    let startIndex = 0;
+    if (lastImagePauseProcessedAdGroupId) {
+      const lastIndex = adGroups.findIndex(
+        adGroup => adGroup.adGroup.id === lastImagePauseProcessedAdGroupId
+      );
+      startIndex = Math.max(lastIndex, 0);
+    }
+    for (let i = startIndex; i < adGroups.length; i++) {
+      const adGroup = adGroups[i];
+      if (this.shouldTerminate()) {
+        Logger.log(
+          `The function is reaching the 6 minute timeout, and therefore will create a trigger to rerun from this ad group: ${adGroup.adGroup.name} and then self terminate.`
+        );
+        PropertiesService.getScriptProperties().setProperty(
+          lastProcessedKey,
+          adGroup.adGroup.id
+        );
+        this.createTriggerForNextRun();
+        return; // Exit the function to prevent further execution
+      }
+      Logger.log(
+        `Processing Ad Group ${adGroup.adGroup.name} (${adGroup.adGroup.id})...`
+      );
+
+      const uploadedImages = this._gcsApi
+        .listImages(this._config['Account ID'], adGroup.adGroup.id, [
+          this._config['Uploaded DIR'],
+        ])
+        ?.items?.map(e => e.name.split('/').slice(-1)[0]);
+
+      if (uploadedImages && uploadedImages.length > 0) {
```

Comment:
I think you could use optional chaining here 
```suggestion
      if (uploadedImages?.length > 0) {
```